### PR TITLE
SOLR-17300: Copy existing listeners on re-creation of Http2SolrClient

### PR DESCRIPTION
**Pull request by iamsanjay**
https://issues.apache.org/jira/browse/SOLR-17300

Instead of #2462, this PR would copy the existing listeners to the new Http2SolrClient. This way re-creating client not requires to call explicitly `Builder#withListenerFactory` to copy existing listeners as `withHttpClient` method automatically  takes care of that. 

Added #2462 test case here.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    try (SolrClient solrClient =
+        cluster.basicSolrClientBuilder().withDefaultCollection(collection).build()) {
+      UpdateRequest commitReq = new UpdateRequest();
+      withBasicAuth(commitReq);
+      for (int i = 0; i < 500; i++) {
+        SolrInputDocument doc = new SolrInputDocument();
+        String[] fields = {"id", i + "", "name", "name = " + i};
+        for (int j = 0; j < fields.length; j += 2) {
+          doc.addField(fields[j], fields[j + 1]);
+        }
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java:80]: This looks clumsy/verbose.  why not `sdoc("id", i, "name", "name = " + i)`

**Review by dsmiley — COMMENTED** [review]
Proposed CHANGES.txt under improvements:

SOLR-17300: Http2SolrClient.Builder.withHttpClient now copies HttpListenerFactory (e.g. for auth, metrics, traces, etc.)

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 2]
```diff
@@ -149,10 +149,6 @@ public void addListenerFactory(HttpListenerFactory factory) {
     this.listenerFactory.add(factory);
   }
 
-  public List<HttpListenerFactory> getListenerFactory() {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:152]: Uh, you are removing a public method of an important SolrJ class?  That should not be done without backwards compatibility considerations.  Leave it please.  Mark it deprecated if you prefer.  The "add" method should be deprecated (for immutability).

File: solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.util.SecurityJson;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class RecoveryZkTestWithAuth extends SolrCloudTestCase {
+  @BeforeClass
+  public static void setupCluster() throws Exception {
+    cluster =
+        configureCluster(2)
```
[the remark this answers is not in the document]
Comment (iamsanjay) [reply solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java]: Yes one node will do. +1 to reduce resources!

[the remark this answers is not in the document]
Comment (epugh) [reply solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java]: Someday in dev-docs, we ought to have "The list of things @dsmiley looks for when writing tests" ;-)

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 2 above)
Comment (dsmiley) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:152]: Ah, very good then

**Comment by dsmiley** [comment]
Hmmm [TestInterleavingTransformer](https://ge.apache.org/scans/tests?search.relativeStartTime=P90D&search.rootProjectNames=solr-root&search.timeZoneId=America%2FNew_York&tests.container=org.apache.solr.ltr.response.transform.TestInterleavingTransformer) failed and it isn't a flapper:

**Comment by iamsanjay** [comment]
> Hmmm [TestInterleavingTransformer](https://ge.apache.org/scans/tests?search.relativeStartTime=P90D&search.rootProjectNames=solr-root&search.timeZoneId=America%2FNew_York&tests.container=org.apache.solr.ltr.response.transform.TestInterleavingTransformer) failed and it isn't a flapper:

Yup it's new. But ran successfully on my machine, tried multiple times now. Should we consider enabling the re-running failing test strategy, or I do not know If it's enabled already or not, to check whether it's flaky or not?  https://docs.gradle.com/develocity/flaky-test-detection/

Evidence To further support it:

# [Flaky Tests at Google and How We Mitigate Them](https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html)
> We even have a way to denote a test as flaky - causing it to report a failure only if it fails 3 times in a row. This reduces false positives, but encourages developers to ignore flakiness in their own tests unless their tests start failing 3 times in a row, which is hardly a perfect solution


# [Cost of Flaky Tests in CI: An Industrial Case Study](https://conf.researchr.org/details/icst-2024/icst-2024-industry/1/Cost-of-Flaky-Tests-in-CI-An-Industrial-Case-Study)
> Automatically rerunning a test costs 0.02~cents, while not rerunning and thus letting the pipeline fail results in a manual investigation costing $5.67 in our context. The insights gained from our case study have led to the decision to shift effort from investigation and repair to automatically rerunning tests

**Pull request merged**